lib/compress/reduce-vars.js

```diff
@@ -528,93 +539,149 @@ function handle_defined_after_hoist(parent) {
+    // Indices of `found_symbols` which are writes
+    const found_symbol_writes = new Set();
+    // Defun ranges are recorded because we don't care if a function uses the def internally
+    const defun_ranges = new Map();
+
+    let tw;
+    parent.walk((tw = new TreeWalker((node, descend) => {
+        if (node instanceof AST_Defun) {
+            const id = node.name.definition().id;
+            if (defuns_of_interest.has(id)) {
                 const start = found_symbols.length;
                 descend();
                 const end = found_symbols.length;
 
-                defun_ranges.set(node, { start, end });
+                defun_ranges.set(id, { start, end });
                 return true;
             }
-            // if we found a defun on the list, mark IN_DEFUN=id and descend
-
-            if (node instanceof AST_Symbol && node.thedef) {
-                const id = node.definition().id;
-                if (symbols_of_interest.has(id)) {
-                    if (node instanceof AST_SymbolDeclaration || is_lhs(node, tw)) {
-                        found_symbol_writes.add(found_symbols.length);
-                    }
-                    found_symbols.push(id);
+        }
+        // if we found a defun on the list, mark IN_DEFUN=id and descend
+
+        if (node instanceof AST_Symbol && node.thedef) {
+            const id = node.definition().id;
+            if (symbols_of_interest.has(id)) {
+                if (node instanceof AST_SymbolDeclaration || is_lhs(node, tw)) {
+                    found_symbol_writes.add(found_symbols.length);
                 }
+                found_symbols.push(id);
             }
-        })));
-
-        for (const { def, defun, fname_def } of potential_conflicts) {
-            const defun_range = defun_ranges.get(defun);
-
-            // find the index in `found_symbols`, with some special rules:
-            const find = (sym_id, starting_at = 0, must_be_write = false) => {
-                let index = starting_at;
-
-                for (;;) {
-                    index = found_symbols.indexOf(sym_id, index);
-
-                    if (index === -1) {
-                        break;
-                    } else if (index >= defun_range.start && index < defun_range.end) {
-                        index = defun_range.end;
-                        continue;
-                    } else if (must_be_write && !found_symbol_writes.has(index)) {
-                        index++;
-                        continue;
-                    } else {
-                        break;
-                    }
-                }
+        }
+    })));
+
+    const defun_first_read_map = new Map();
+    for (const defun of defuns_of_interest) {
+        const defun_range = defun_ranges.get(defun);
+        const defun_first_read = found_symbols.findIndex(
+            (sym, index) =>
+                sym === defun &&
+                (index < defun_range.start || index > defun_range.end)
```

Comment:
```suggestion
                (index <= defun_range.start || index > defun_range.end)
```

Looks like an off-by one error. `start` is the index of the first symbol inside the function, `end` is the index+1 of the last symbol.